### ISSUE TITLE
[FEA] Multi-Column groupby aggregations when grouping by string columns (multi-index)

### ISSUE THREAD

**Issue by ayushdg** [rapidsai/cudf#1570]
This issue is a way to keep track of improvements and/or additions that can be made to MultiIndex support in cudf. @thomcom

When trying attempting aggregations on multi-column groupby where one of the columns being grouped by is a string column, the operation fails.

```
import cudf
import numpy as np

df = cudf.DataFrame()
df['a'] = ['a','b','c']*3
df['b'] = [1,2,3,2,2,3,3,1,2]
df['c'] = np.random.randint(100,size=9)
df.groupby(by=['a','b']).min()
```
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-22-153e55632bb5> in <module>
      3 df['b'] = [1,2,3,2,2,3,3,1,2]
      4 df['c'] = np.random.randint(100,size=9)
----> 5 df.groupby(by=['a','b']).min().to_pandas()

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.3.0+4024.g5ed8a94.dirty-py3.7-linux-x86_64.egg/cudf/groupby/groupby.py in min(self, sort)
    200 
    201     def min(self, sort=True):
--> 202         return self._apply_basic_agg("min", sort)
    203 
    204     def max(self, sort=True):

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.3.0+4024.g5ed8a94.dirty-py3.7-linux-x86_64.egg/cudf/groupby/groupby.py in _apply_basic_agg(self, agg_type, sort_results)
    111             The aggregation function to run.
    112         """
--> 113         return _cpp_apply_basic_agg(self, agg_type, sort_results=sort_results)
    114 
    115     def apply_multiindex_or_single_index(self, result):

cudf/bindings/groupby.pyx in cudf.bindings.groupby._apply_basic_agg()

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.3.0+4024.g5ed8a94.dirty-py3.7-linux-x86_64.egg/cudf/groupby/groupby.py in apply_multiindex_or_single_index(self, result)
    132             # time the groupby is calculated.
    133             for by in self._by:
--> 134                 level = result[by].unique()
    135                 code = result[by]
    136                 for idx, value in enumerate(level):

/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.3.0+4024.g5ed8a94.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/series.py in unique(self, method, sort)
   1210         if self.null_count == len(self):
   1211             return np.empty(0, dtype=self.dtype)
-> 1212         res = self._column.unique(method=method)
   1213         return Series(res)
   1214 

AttributeError: 'StringColumn' object has no attribute 'unique'
```

**Comment by thomcom** [comment]
@mrocklin this feature is important for dask-cudf in 0.7 isn't it?

**Comment by mrocklin** [comment]
I don't know.  Can you explain more about why this is important for dask-cudf for 0.7?  Sorry I may be a bit out of it.

**Comment by thomcom** [comment]
It is a groupby with a StringColumn - so grouping by strings instead of numeric values. We added strings in 0.6 and better groupby in 0.7, but the StringColumn doesn't have a `unique` method yet, so we can't group by strings. Do you remember if it is important for 0.7?

**Comment by mrocklin** [comment]
I genuinely don't recall.  Sorry, I've been out of it handling UCX and other things recently.  I would ask @kkraus14

**Comment by ayushdg** [comment]
@thomcom I can implement a `unique` method for StringColumns (it's a simple enough addition). The groupby would still be blocked by the the lack of binops like an `==` or working nicely with `masked_assign`
https://github.com/rapidsai/cudf/blob/branch-0.7/python/cudf/groupby/groupby.py#L187-L188

**Comment by kkraus14** [comment]
@ayushdg I believe this should be resolved in 0.7, can you confirm?

**Comment by ayushdg** [comment]
@kkraus14 Pr #1701 was merged in 0.8 which I believe is needed to groupby string columns. Would you like me to test with 0.7 or 0.8?

**Comment by kkraus14** [comment]
I believe 0.7 was updated to not need to call `.unique` on the MultiIndex for groupbys as well, so give both a shot.

**Comment by ayushdg** [comment]
Can confirm that I can perform the groupby with `0.7.1` and I get a resulting Dataframe. However there is no way to actually view the output as `to_pandas` or `head` is still dependent on the `unique` method. 

On `0.8` can confirm do not see the issue of missing `unique` method. However we might still be lacking more method implementations:
```
>>> df.groupby(by=['a','b']).min().to_pandas()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.1+312.gc760343.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/dataframe.py", line 2299, in to_pandas
    index = self.index.to_pandas()
  File "/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.1+312.gc760343.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/multiindex.py", line 362, in to_pandas
    for code in self.codes.columns:
  File "/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.1+312.gc760343.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/multiindex.py", line 138, in codes
    self._compute_levels_and_codes()
  File "/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.1+312.gc760343.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/multiindex.py", line 168, in _compute_levels_and_codes
    level, Series(range(len(level))))
  File "/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.1+312.gc760343.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/series.py", line 993, in replace
    result = self._column.find_and_replace(to_replace, value)
  File "/conda/envs/rapids/lib/python3.7/site-packages/cudf-0.7.1+312.gc760343.dirty-py3.7-linux-x86_64.egg/cudf/dataframe/columnops.py", line 90, in find_and_replace
    raise NotImplementedError
NotImplementedError
```

@kkraus14  Can leave this issue open or can close this one and open a new issue specific to the stacktrace pasted above.

**Comment by thomcom** [comment]
There is now a `unique` method in StringColumn. I'm not sure if it is in 0.7.2 or 0.8.0. There is an issue with grouping by stringcolumns, however, as `gdf_find_and_replace_all` does not work with them. I'll file an issue on that.

**Comment by ayushdg** [comment]
Hey @thomcom `unique` was merged for 0.8. It would be great if you create a new issue for `gdf_find_and_replace_all`. I'll keep this issue open and retest the code when there is a feature for `gdf_find_and_replace_all` with string columns and keep everyone updated if anything new comes up.

**Comment by kkraus14** [comment]
@ayushdg we refactored out the find and replace calls with joins and this now works as expected. Closing.
